torch/csrc/jit/testing/file_check.cpp

```diff
@@ -0,0 +1,329 @@
+//==-- llvm/Support/FileCheck.h ---------------------------*- C++ -*-==//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+// API modified from llvm::FileCheck
+
+#include <c10/util/Exception.h>
+#include <c10/util/Optional.h>
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/csrc/jit/source_range.h>
+#include <algorithm>
+#include <iostream>
+#include <sstream>
+#include <string>
+
+#include <torch/csrc/jit/passes/python_print.h>
+#include <torch/csrc/jit/testing/file_check.h>
+
+namespace torch {
+namespace jit {
+
+void printQuotedString(std::ostream& stmt, const std::string& str);
+
+namespace testing {
+
+enum CheckType {
+  CHECK,
+  CHECK_NEXT,
+  CHECK_SAME,
+  CHECK_NOT,
+  CHECK_COUNT,
+  CHECK_DAG,
+};
+
+struct Check {
+  Check(
+      CheckType type,
+      std::string str,
+      c10::optional<size_t> count = c10::nullopt)
+      : type_(type), search_str_(std::move(str)) {
+    count_ = std::move(count);
+  };
+
+  CheckType type_;
+  c10::optional<size_t> count_;
+  const std::string search_str_;
+
+  friend std::ostream& operator<<(std::ostream& out, const Check& c);
+};
+
+std::ostream& operator<<(std::ostream& out, const Check& c) {
+  switch (c.type_) {
+    case CHECK:
+      out << "CHECK";
+      break;
+    case CHECK_NEXT:
+      out << "CHECK-NEXT";
+      break;
+    case CHECK_SAME:
+      out << "CHECK-SAME";
+      break;
+    case CHECK_NOT:
+      out << "CHECK-NOT";
+      break;
+    case CHECK_DAG:
+      out << "CHECK-DAG";
+      break;
+    case CHECK_COUNT:
+      out << "CHECK-COUNT-" << *c.count_;
+      break;
+  }
+  out << ": " << c.search_str_;
+  return out;
+};
+
+namespace {
+size_t assertFind(
+    const SourceRange& search_range,
+    const std::string& sub,
+    const Check& check) {
+  auto pos = search_range.file().find(sub, search_range.start());
+  if (pos == std::string::npos || (pos + sub.size()) > search_range.end()) {
+    auto range = SourceRange(
+        std::make_shared<std::string>(search_range.file()),
```

Comment:
This is copying the string. You can use `file_ptr` to get the shared pointer to the common string.